### ISSUE TITLE
SCP session stuck after file transfer

### ISSUE THREAD

**Issue by oddko** [ssh-mitm/ssh-mitm#100]
Hi,

Not sure if this is an issue with paramiko or this utility, but SCP transfers seem to hang once the files got transferred (requiring a good ol CTRL+C to end). From what I can gather from reading the code of the SCP forwarders is that they have no way to tell how many files are being transfered and so just keep waiting for more once the last file being transfered returned an exit code of 0. This makes it very obvious that a mitm attack is occuring since SCP commands usually exit gracefully (With the client ending the connection with a TCP RST ACK). Here is where it ends (stuck) when transferring to the SSH server of a raspberry pi (with a fresh install) from Windows (that itself runs ssh-mitm):

![log](https://user-images.githubusercontent.com/15215305/163038682-d97c5297-237c-49ed-a2b6-c680bf5eb5ab.png)

**Comment by oddko** [comment]
Some comparison (using debug level 3 on the client SCP) indicates it remains stuck once the client SCP sent an EOF (indicating it has sent all files). Here's a comparison of the log output between using scp through the mitm utility vs connecting to the real server directly:

<img width="1190" alt="Capture d’écran 2022-04-15 à 09 49 15" src="https://user-images.githubusercontent.com/15215305/163541773-0ead1e3d-db34-47d4-8dec-925aa511e36c.png">

**Comment by manfred-kaiser** [comment]
Thanks for the bug report and the analysis.

I will try to fix it in the next weeks.

With the release of OpenSSH 9 the command "scp" uses the "sftp protocol" background and some a new remote copy command for sftp was introduced. So I have to test the scp and sftp functionality also against the new OpenSSH release. https://www.openssh.com/releasenotes.html

So it could take a bit longer to fix this bug.

**Comment by manfred-kaiser** [comment]
I can't reproduce this error with SSH-MITM 1.0.0 and 2.0.2 in combination with OpenSSH 8.9 (latest OpenSSH version)

Which version of SSH-MITM and OpenSSH are you using?

Can you provide the command line options which are used to start SSH-MITM and to copy files with SCP?

**Comment by manfred-kaiser** [comment]
Ok, I found the error. I was only testing copy operations from server to client, but not client to server as you did - I should read the whole issue tickets ;-)

Steps to reproduce:

Start SSH-MITM:
```
ssh-mitm --debug --paramiko-log-level debug server
```

I got the EOF, but as you mentioned, the forwarder does not handle the EOF.

```
[05/22/22 15:06:04] DEBUG    [chan 0] EOF received (0)                                                            channel.py:1212
```

Thanks for the detailed bug report.

**Comment by manfred-kaiser** [comment]
Issue fixed with commit df46c079c5f089a80572d06cb1d39c41225bb8b2

**Comment by oddko** [comment]
Hi, thanks for the fix ! It indeed now closes the session, though SCP returns with a non-zero exit code so it might not be exiting that gracefully, I'll try investigate.

**Comment by manfred-kaiser** [comment]
I will try to fix this issue with the next release. Thanks for testing.

**Comment by oddko** [comment]
Hi,

Unfortunately no, still the exact same behavior as posted previously, goes straight for a session close. Given that it is the client that starts "terminating" the session through an EOF wouldn't it be the that code https://github.com/ssh-mitm/ssh-mitm/blob/develop/sshmitm/forwarders/scp.py#L95-L97 that handles it (and as it currently stands, closes the session immediately) ? Still client to server operations btw.

**Comment by zoey-fux** [comment]
the provided patch should fix the mentioned problem

before closing the session, exit_status 0 will be sent to the client.

I'm not sure if it's possible to reorder the messages to match OpenSSH, but I think this should not be a big problem.

I have added a screenshot of the scp client with "SSH-MITM" and OpenSSH 8.9.

@oddko can you check if the provided fix in my pull request #105 solved the problem?

![exit_status_0](https://user-images.githubusercontent.com/103361677/173766288-1d005b2f-64f4-43f7-a3a4-28963b624ce2.png)

**Comment by zoey-fux** [comment]
I have created a new fix which sends eof before sending the return code. Now it should match better the behavior of OpenSSH.

![send eor before sending return code](https://user-images.githubusercontent.com/103361677/174236116-61899cc1-1774-43f3-8759-12877949fcf5.png)
